src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -627,468 +719,378 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
+    {
+        auto sleep_before_commit_local_part_in_replicated_table_ms = storage.getSettings()->sleep_before_commit_local_part_in_replicated_table_ms;
+        if (sleep_before_commit_local_part_in_replicated_table_ms.totalMilliseconds())
+        {
+            LOG_INFO(log, "committing part {}, triggered sleep_before_commit_local_part_in_replicated_table_ms {}",
+                     part->name, sleep_before_commit_local_part_in_replicated_table_ms.totalMilliseconds());
+            sleepForMilliseconds(sleep_before_commit_local_part_in_replicated_table_ms.totalMilliseconds());
         }
-        else
-            throw Exception(ErrorCodes::LOGICAL_ERROR,
-                            "Conflict block ids and block number lock should not "
-                            "be empty at the same time for async inserts");
+    };
 
-        if constexpr (!async_insert)
-        {
-            if (!existing_part_name.empty())
-            {
-                LOG_DEBUG(log, "Will check part {} checksums", existing_part_name);
-                try
-                {
-                    NameSet unused;
-                    /// if we found part in deduplication hashes part must exists on some replica
-                    storage.checkPartChecksumsAndAddCommitOps(zookeeper, part, ops, existing_part_name, unused);
+    auto commit_new_part_stage = [&] () -> CommitRetryContext::Stages
+    {
+        chassert(retry_context.block_number_lock.has_value());
 
-                    /// We have to check that the block_id still exists to avoid a race condition with DROP_RANGE
-                    block_unlock_op_idx = ops.size();
-                    ops.emplace_back(zkutil::makeCheckRequest(storage.zookeeper_path + "/blocks/" + block_id, -1));
-                }
-                catch (const zkutil::KeeperException &)
-                {
-                    throw;
-                }
-                catch (const Exception & ex)
-                {
-                    if (ex.code() == ErrorCodes::CHECKSUM_DOESNT_MATCH)
-                    {
-                        LOG_INFO(
-                            log,
-                            "Block with ID {} has the same deduplication hash as other part {} on other replica, but checksums (which "
-                            "include metadata files like columns.txt) doesn't match, will not write it locally",
-                            block_id,
-                            existing_part_name);
-                        return;
-                    }
-                    throw;
-                }
-            }
-            else
-            {
-                /// Information about the part.
-                storage.getCommitPartOps(ops, part, block_id_path);
-            }
-        }
-        else
+        /// Lock might be easily lost at some retry attempt due to connection loss
+        /// Redo the locking
+        if (!retry_context.block_number_lock->isLocked())
```

Comment:
That doesn't work.
When error injected, saved ephemeral nodes are not expired. They are gone, but